File: 0649-dota2-senate/solution.java
class Solution {
    public String predictPartyVictory(String senate) {
        int n = senate.length();
        Queue<Integer> radiant = new LinkedList<>();
        Queue<Integer> dire = new LinkedList<>();
        
        // 1. חלוקת הסנטורים לתורים לפי המיקום המקורי שלהם
        for (int i = 0; i < n; i++) {
            if (senate.charAt(i) == 'R') {
                radiant.add(i);
            } else {
                dire.add(i);
            }
        }
        
        // 2. סימולציית ההצבעה
        while (!radiant.isEmpty() && !dire.isEmpty()) {
            int rIdx = radiant.poll();
            int dIdx = dire.poll();
            
            // הסנטור עם האינדקס הנמוך יותר מצביע קודם ומבטל את השני
            if (rIdx < dIdx) {
                // Radiant מנצח ומצטרף לסוף התור לסיבוב הבא
                radiant.add(rIdx + n);
            } else {
                // Dire מנצח ומצטרף לסוף התור לסיבוב הבא
                dire.add(dIdx + n);
            }
        }
        
        // 3. המנצח הוא זה שנשאר לו סנטור בתור
        return radiant.isEmpty() ? "Dire" : "Radiant";
    }
}

File: 0649-dota2-senate/solution.py
from collections import deque

class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        queue_R = deque()
        queue_D = deque()
        n = len(senate)
        
        # 1. Populate the queues with the indices of each party's senators
        for i, party in enumerate(senate):
            if party == 'R':
                queue_R.append(i)
            else:
                queue_D.append(i)
                
        # 2. Simulate the voting rounds until one queue is empty
        while queue_R and queue_D:
            r_index = queue_R.popleft()
            d_index = queue_D.popleft()
            
            # The senator with the smaller index acts first and bans the other
            # The survivor goes back to the queue for the next round (index + n)
            if r_index < d_index:
                queue_R.append(r_index + n)
            else:
                queue_D.append(d_index + n)
                
        # 3. Whichever queue still has senators remaining wins
        return "Radiant" if queue_R else "Dire"
